### ISSUE TITLE
[feature request] Cache-control: Immutable

### ISSUE THREAD

**Issue by aspnet-hello** [dotnet/aspnetcore#2611]
_From @tinchou on Sunday, January 29, 2017 12:17:57 PM_

This is a very recent feature, but one that could improve many sites performance. If there already exists a way to achieve this, it would be great to have it on the Docs.

Info: https://code.facebook.com/posts/557147474482256, https://hacks.mozilla.org/2017/01/using-immutable-caching-to-speed-up-the-web/

Thanks!

_Copied from original issue: aspnet/ResponseCaching#97_

**Comment by aspnet-hello** [comment]
_From @guardrex on Sunday, January 29, 2017 3:42:05 PM_

> it would be great to have it on the Docs

It's not currently documented because `OnPrepareResponse` isn't in the static files doc. I opened an issue to discuss it over there. https://github.com/aspnet/Docs/issues/2625

Regarding the feature request: Is your request to add something to the `ResponseCache` attribute for this? If so, wouldn't that be an MVC feature request? It doesn't seem like the middleware here would have anything to do with this directive.

**Comment by aspnet-hello** [comment]
_From @guardrex on Sunday, January 29, 2017 3:59:44 PM_

This middleware reads the header to determine if and how long to cache a resource. It doesn't set it.

Speaking of setting it, yes, either it's something for you to do with Static File Middleware (i.e., in `OnPrepareResponse`), or you could make an argument that something should be added to the `ResponseCache` attribute bits that would allow you to configure it easily for MVC-based responses. The attribute is an MVC thing, so the request would have to be made over in the MVC repo. @JunTaoLuo will confirm or correct me on these points.

**Comment by aspnet-hello** [comment]
_From @Tratcher on Monday, January 30, 2017 9:55:36 AM_

Interesting feature. This middleware is not the right place for it though, it doesn't set any headers, it only reads them. Using OnPrepareResponse sounds like your best place to implement this. Note we've previously declined requests to add other cache control options directly to the static files options, so I doubt we'd add this one as a first class option, especially before it becomes a formal spec.

Adding this option to the MVC attribute seems out of place, as MVC content is inherently dynamic. It should be pretty easy to append this value to the header in your MVC controller if you really want it. Give it a try and feel free to share the code for review.

**Comment by aspnet-hello** [comment]
_From @RehanSaeed on Thursday, October 19, 2017 1:22:54 AM_

I think the `HttpContext` could use the following extension method being added to it:

```
public static HttpContext ApplyCacheProfile(this HttpContext context, CacheProfile cacheProfile)
```

The `CacheProfile` class also needs an `IsImmutable` boolean flag on it. I've added it to my Boilerplate.AspNetCore NuGet package [here](https://github.com/ASP-NET-Core-Boilerplate/Framework/blob/118c3e00cc311b435d33856f42b3a8a148fcb935/Source/Boilerplate.AspNetCore/HttpContextExtensions.cs#L29) if anyone wants to see the code. I think I remember @andrewlock had raised a GitHub issue to get this method added but the issue was closed.
